### PR TITLE
automation testing changes added

### PR DESCRIPTION
**Pull request by jdcanas**
new and improved

Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/Notebook.scala
Comment (Qi77Qi) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/Notebook.scala]: Use sealed trait

```
sealed trait NotebookMode extends Product with Serializable {
   def asString: String
}
object NotebookMode {
    final case object SafeMode extends NotebookMode {
    def asString: String = "playground"
   }

    final case object EditMode extends NotebookMode {
     def asString: String = "edit"
   }

   val stringToMode: Map[String, NotebookMode] = List(SafeMode, EditMode).map(x => x.asString -> x).toMap
}
```

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookPage.scala
Comment (Qi77Qi) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookPage.scala]: elementIds
      .map(elementId => find(id(elementId)).exists(_.underlying.getAttribute("style") == "display: none;"))
      .forall(identity)

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookTestUtils.scala
Comment (Qi77Qi) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookTestUtils.scala:326]: formatting

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookTestUtils.scala
Comment (Qi77Qi) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookTestUtils.scala]: google2 has a https://github.com/broadinstitute/workbench-libs/blob/develop/google2/src/main/scala/org/broadinstitute/dsde/workbench/google2/GoogleStorageInterpreter.scala#L80 that allows you to download an object to a local file

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/Welder.scala
Comment (Qi77Qi) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/Welder.scala]: remove println

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/JupyterModel.scala
Comment (Qi77Qi) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/JupyterModel.scala:30]: `created` and `last_modified` should ideally be `Instant`

`path` can be `java.nio.file.Path`

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterFixtureSpec.scala
Comment (Qi77Qi) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterFixtureSpec.scala]: Prefer to use `Some` since there's no way `MachineConfig` can be null. It's better to be explicit about intent (using Option implies to me what you're wrapping around can be null)

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/JupyterModel.scala
Comment (jdcanas) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/JupyterModel.scala:30]: I'm maintaining parity with the existing `ContentItem` test class. These classes are used to serialize an API response, so the simpler the types the better as I see it.

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/Notebook.scala
Comment (jdcanas) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/Notebook.scala]: updated to sealed trait, but had to implement stringToMode slightly differently, as the string we are converting comes from the UI and lacks type safety

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterFixtureSpec.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterFixtureSpec.scala:30]: Should this comment be moved up?

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterFixtureSpec.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterFixtureSpec.scala]: These can be vals because they're intended to be immutable. The other fields are vars because they are set in test setup code, which is not the case here.

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterFixtureSpec.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterFixtureSpec.scala]: Nice, so existing tests will default to having welder enabled?

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterFixtureSpec.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterFixtureSpec.scala]: Add these back

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/JupyterModel.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/JupyterModel.scala]: Is this class used? If not, remove it.

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/JupyterModel.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/JupyterModel.scala:16]: Is this comment still true? I see below you are deserializing `NotebookContentItem` from the Jupyter API.

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookExtensionSpec.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookExtensionSpec.scala]: Reminder to remove debug mode before merge

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookExtensionSpec.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookExtensionSpec.scala]: Might as well keep this test, no?

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookTestUtils.scala
Comment (rtitle) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookTestUtils.scala:326]: indentation seems off

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookExtensionSpec.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookExtensionSpec.scala]: Remove this :)

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookExtensionSpec.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookExtensionSpec.scala]: Uncomment

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookPage.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookPage.scala:175]: This function does not wait for the cell to return though -- but I agree it does not have the above bug. Maybe note the difference in a comment.

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookTestUtils.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookTestUtils.scala]: Remove this method?

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterFixtureSpec.scala
Comment (rtitle) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterFixtureSpec.scala]: Answering my own question: it seems like tests inheriting `ClusterFixtureSpec` will have welder enabled, but some tests use `LeonardoTestUtils.createNewCluster()` directly. I wonder if we should enable welder for those tests too. It would be good to make sure welder doesn't affect any other functionality..

**Review by rtitle — CHANGES_REQUESTED** [review]
Some cleanup is needed, but overall looks good.

It would be good to run the full suite of tests with welder enabled, to make sure welder doesn't cause any regressions.

**Comment by jdcanas** [comment]
This is in a non-reviewable state (with some clean-up needed) because the branch is being used to attempt to diagnose https://broadworkbench.atlassian.net/browse/IA-1153

File: .gitignore
Comment (Qi77Qi) [inline .gitignore:32]: I would do `automation/output/`

**Comment by jdcanas** [comment]
jenkins retest

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookExtensionSpec.scala
Comment (Qi77Qi) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookExtensionSpec.scala:92]: it shouldn't take that long 🤔

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookExtensionSpec.scala
Comment (Qi77Qi) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/notebooks/NotebookExtensionSpec.scala]: I would rename `googleCloudDir` to `gcsPath` since it's a gcs path to an object, not just a directory

**Comment by jdcanas** [comment]
jenkins retest

**Comment by jdcanas** [comment]
jenkins retest

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/JupyterModel.scala
Comment (Qi77Qi) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/JupyterModel.scala:23]: `size` is the only field being used currently. Talked to Justin offline and he'll either remove them in this PR or a followup PR

**Review by Qi77Qi — APPROVED** [review]
few more comments (especially https://github.com/DataBiosphere/leonardo/pull/1002/files#diff-fed0726a11a83b9a5b2720ed121689f5R92)...approving assuming they'll be address in a follow up PR since there's bug fix for UI mixed in this PR

File: automation/project/Dependencies.scala
Comment (rtitle) [inline automation/project/Dependencies.scala:62]: Just curious what we're using guava for.

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterFixtureSpec.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterFixtureSpec.scala:33]: I would make these `val`s since they are not expected to be mutated.

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/JupyterModel.scala
Comment (rtitle) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/JupyterModel.scala:23]: Agreed about only defining the fields we're using. The json mapper should ignore unknown fields.

**Review by rtitle — APPROVED** [review]
A couple minor things above but overall 👍 from me. OK to merge - minor things could be addressed in a followup PR.

File: automation/project/Dependencies.scala
Comment (jdcanas) [reply automation/project/Dependencies.scala:62]: needed for google2 update

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/JupyterModel.scala
Comment (jdcanas) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/JupyterModel.scala:23]: will remove with subsequent automation tests

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterFixtureSpec.scala
Comment (jdcanas) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/ClusterFixtureSpec.scala:33]: they are mutated on a per-spec basis based on which specs would like debug mode

File: automation/project/Dependencies.scala
Comment (Qi77Qi) [reply automation/project/Dependencies.scala:62]: it's not needed explicitly anywhere. But it's used by several google related libraries. Forcing the version here to solve an issue somehow a lower version is brought in which lead to compile error.

**Pull request merged**
